### PR TITLE
feat: Effect Handlers

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

**Comment by JonathanStarup** [comment]
!!!jar

**Comment by JonathanStarup** [comment]
Okay it seems i was presumptous

**Comment by JonathanStarup** [comment]
!!!jar

**Comment by JonathanStarup** [comment]
@magnus-madsen effects now work! (75% of tests pass, the rest fail because of unresolved erasure issues)

**Comment by magnus-madsen** [comment]
> @magnus-madsen effects now work! (90% of tests pass, the rest fail because of unresolved erasure issues)

Nice! You mean as exceptions or with resumption?

What about the unresolved erasure issues?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala [hunk 1]
```diff
@@ -61,7 +61,7 @@ sealed trait BackendObjType {
     case BackendObjType.Arrays => JvmName(JavaUtil, "Arrays")
     case BackendObjType.StringBuilder => JvmName(JavaLang, "StringBuilder")
     case BackendObjType.Objects => JvmName(JavaLang, "Objects")
-    case BackendObjType.LambdaMetaFactory => JvmName(JavaLangInvoke, "LambdaMetaFactory")
+    case BackendObjType.LambdaMetaFactory => JvmName(JavaLangInvoke, "LambdaMetafactory")
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala:64]: 🤦  😆

**Comment by magnus-madsen** [comment]
Not really sure where the errors are now.

**Comment by JonathanStarup** [comment]
> > @magnus-madsen effects now work! (90% of tests pass, the rest fail because of unresolved erasure issues)
> 
> Nice! You mean as exceptions or with resumption?
> 
> What about the unresolved erasure issues?

All effect functionality should work now!

The issue is the conversion where all functions return a Boxed value. Its pretty ad hoc and i think ive done it in a wrong way. in this pr everything is type a -> Obj, but really i should just keep the a -> b type and make that mean "a function that takes a and returns a value which you can extract b from"

**Comment by magnus-madsen** [comment]
Do you have an example? I tried this:

```
eff Throw {
    pub def throw(): Unit
}

def main(): Unit \ IO =
    let r = f();
    println("hello: ${r}")


def f(): String  =
    try {
        f()
    } with Throw {
        def throw(_k) = "a"
    }

def g(): Unit \ Throw = do Throw.throw()

```

**Comment by magnus-madsen** [comment]
I guess we need to lift
```
-- Syntax Error -------------------------------------------------- foo.flix

>> Non-Unit return type. All effect operations must return Unit.

2 |     pub def throw(): Int32
                ^^^^^
                non-Unit return type
```

to be able to write useful programs.

**Comment by magnus-madsen** [comment]
```

**Comment by JonathanStarup** [comment]
Yeah, I've discussed it with Matt. We should draft some type rules and implement them

**Comment by JonathanStarup** [comment]
@magnus-madsen The only thing broken now should be effect operations that return non-object return types (will fix that now in the eraser)

**Comment by magnus-madsen** [comment]
> @magnus-madsen The only thing broken now should be effect operations that return non-object return types (will fix that now in the eraser)

And a conflict.

**Comment by magnus-madsen** [comment]
Are you marking this as RFR?

File: main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala [hunk 2]
```diff
@@ -294,5 +327,68 @@ object GenFunAndClosureClasses {
     m.visitEnd()
   }
 
+  private def compileOnCall(v: ClassWriter, classType: JvmType.Reference, defn: Def): Unit = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala:330]: Is this entire method for debugging?

**Comment by JonathanStarup** [comment]
@magnus-madsen not polished for merging yet, but ready for testing :) no known bugs remain

**Pull request merged**